command_line/expeditor.py

```diff
@@ -0,0 +1,213 @@
+
+    return scans
+
+
+def reflections_fill_scan(reflections, scan, step_degrees):
+    """Verify that the reflections defined in the input have Z centroids which
+    fill the scan."""
+
+    z = reflections["xyzobs.px.value"].parts()[2]
+
+    i0, i1 = scan.get_array_range()
+    slots = int(round((i1 - i0) * scan.get_oscillation()[1] / step_degrees))
+
+    fill = flex.histogram(z, data_min=i0, data_max=i1, n_slots=slots)
+
+    for s in fill.slots():
+        if s == 0:
+            return False
+
+    return True
+
+
+class Protocol(object):
+    def __init__(self):
+        usage = "usage: dials.expeditor [options] indexed.expt indexed.refl"
+
+        self.parser = OptionParser(
+            usage=usage,
+            phil=phil_scope,
+            read_reflections=True,
+            read_experiments=True,
+            check_format=False,
+            epilog=help_message,
+        )
+
+    def run(self):
+        params, _ = self.parser.parse_args(show_diff_phil=True)
+
+        if not params.input.experiments:
+            print("No experiments found in the input")
+            self.parser.print_help()
+            return
+        if not params.input.reflections:
+            print("No reflections found in the input")
+            self.parser.print_help()
+            return
+
+        if len(params.input.reflections) != len(params.input.experiments):
+            print("Cardinality error")
+            self.parser.print_help()
+            return
+
+        experiments = flatten_experiments(params.input.experiments)
+        reflections = flatten_reflections(params.input.reflections)[0]
+
+        print(model_connectivity(experiments))
+        print()
+
+        z = reflections["xyzobs.px.value"].parts()[2]
+
+        experiments_out = ExperimentList()
+        reflections_id = flex.int(reflections.size(), -424242)
+        reflections_id.set_selected(reflections["id"] == -1, -1)
+
+        crystal_scan = {}
+
+        for j, e in enumerate(experiments):
+            if e.crystal in crystal_scan:
+                if e.scan is not crystal_scan[e.crystal]:
+                    e.crystal = copy.deepcopy(e.crystal)
+            else:
+                crystal_scan[e.crystal] = e.scan
+
+            sel = reflections.select(reflections["id"] == j)
+            i0, i1 = e.scan.get_image_range()
+            nref = sel.size()
+            scans = select_scans_from_reflections(sel, e.scan)
+
+            assert len(scans) > 0
+
+            eid = len(experiments_out)
+
+            e.scan = scans[0]
+            experiments_out.append(e)
+
+            # rewrite id
+            i0, i1 = e.scan.get_array_range()
+            sel = (reflections["id"] == j) & (z >= i0) & (z <= i1)
+            reflections_id.set_selected(sel, eid)
+
+            print(
+                "Output experiment %d has %d / %d reflections"
+                % (eid, sel.count(True), nref)
+            )
+
+            for k, s in enumerate(scans[1:]):
```

Comment:
I don't see why this can't simply be 
```suggestion
            for k, s in enumerate(scans):
```
and eliminate the need for the duplicated code just above dealing with the first scan.